README.md
# Epicopy

Epicopy is an R package that obtains copy number variation (CNV) information from Illumina Human Methylation 450K microarrays. If the users have their own reference normal samples or decide to not use any normals for reference (see Usage section below), Epicopy can be used as a standalone package.

If not, users may choose to install its companion package **"EpicopyData"** which contains a series of normal tissues compiled from the Cancer Genome Atlas (TCGA) effort. It contains normal tissues from three sources; thyroid, breast, and lung. For more information on EpicopyData please visit the [EpicopyData page](https://github.com/sean-cho/EpicopyData).

---

## Installation

The preferred method for installation is through Hadley Wickham's `install_github` function through the package [devtools](https://github.com/hadley/devtools).

If users have their own normal samples, it is recommended that they install the standalone version as it is much smaller in size.

### Standalone

Do not build the vignette as EpicopyData is a dependency.

To install the standalone package use the following line:
```
devtools::install_github('sean-cho/Epicopy')
```

### Complete

**Warning:** The EpicopyData file contains a large data file and requires the manual installation of EpicopyData.

The complete version includes all the normals compiled from TCGA.

#### Step 1: Install EpicopyData

Please download the binary release [here](https://github.com/sean-cho/EpicopyData/releases/download/v1.0.1/EpicopyData_1.0.1.tar.gz) and run the following lines of code.

```
path <- "path.to.file/EpicopyData_1.0.1.tar.gz"
install.packages(path, repos = NULL, type = 'source')
```

#### Step 2a: Install Epicopy without vignette

```
devtools::install_github('sean-cho/Epicopy')
```

#### Step 2b: Install Epicopy with vignette
**Warning:** Please note that the vignette includes the complete processing of three example raw data files and may take a while to build.

```
devtools::install_github('sean-cho/Epicopy', build_vignette = TRUE)
```

---

## Usage

### Standalone

Since no vignette is built, users can view the online vignette for examples [here](https://github.com/sean-cho/Epicopy/blob/master/vignettes/Epicopy.Rmd). Additional information is also available in the help files.

The key to using Epicopy as a standalone package is to specify normals in the samplesheet or use the argument `Normals = NA` in the `epicopy` or `getLRR` functions. This is important, as the default argument uses all normals included in the EpicopyData package.

### With EpicopyData normals

Install and build the vignette. Users may specify any of the normals included in with the EpicopyData package using the options:
- `all` or `NULL` (default): Uses all EpicopyData normals.
- `thyroid`, `breast`, or `lung`: Uses normal tissue of specified organ.

More information is available using `?getLRR`.


inst/NEWS
Initial release to Bioconductor.

R/getLRR.R
#' Obtain LRR from RGset
#'
#' \code{getLRR} Preprocess RGset and obtain LRR for samples against a 
#'    reference set.
#'  
#' @description Returns a matrix containing the log R ratios (LRR) of samples 
#'    to reference set. Input file has to be an RGset. Normals can be defined 
#'    by the user (see details).
#' 
#' @param rgSet RGChannelSet object containing samples of interest
#' @param Normals Either (1) RGChannelSet of normals (reference set), (2) a 
#'    logical or numeric vector  denoting the normals in the input rgSet, (3) a
#'    character of 'thyroid', 'lung', 'breast', or 'all' denoting reference 
#'    set included in Epicopy, or (4) defaults to NULL, which uses all normals 
#'    in Epicopy.
#' @param sampNames A character vector giving sample names. Defaults to NULL, 
#'    which uses \code{pData(rgSet)$Sample_Name}.
#' @param QN Logical. Perform probe type specific quantile normalization of 
#'    reference set.
#' @param Ref 'mode' or 'median'. Method to calculate reference intensity. 
#'    Defaults to mode using the modeest package.
#' @param mode.bw If Ref = 'mode', uses this bw to estimate the mode. 
#'    Defaults to 0.1.
#' @param mode.method If Ref = 'mode', uses this method to estimate 
#'    the mode. Defaults to 'naive'.
#' @param normal.cnv Logical. Return LRR for reference set.
#' @param mean.center Logical. Perform mean centering of final data.
#' @param filterProbes Logical. Filter probe by TCGA defined 450K probes 
#'    adjacent to a known SNP.
#' @param retainedProbes Character. IF filterProbes is TRUE, denotes probes to keep. 
#'    If NULL, uses probes retained by TCGA.
#' @param keep_fnobj Logical. Save functional normalized object as rda.
#' @param fn_output Directory for fnobject. Defaults to current directory.
#' @param ... Passes argument to preprocessFunnorm.
#' 
#' @details GetLRR takes RGset as an argument as an input.
#' 
#' Normals/Reference samples can be defined multiple ways. Defaults to 
#' NULL which uses all the normals incorporated by the Epicopy package 
#' (see next paragraph). Providing an RGset will BiocGenerics::combine(rgSet, Normals) to 
#' get a dataset. Providing a logical/numeric vector will subset the rgSet 
#' into query and reference, with TRUE/numbers representing reference samples. 
#' Users can also specify one of 'thyroid', 'breast', 'lung', or 'all' to use 
#' Epicopy normals.
#' 
#' Epicopy normals are compiled using solid normal tissue methylation microarrays 
#' from TCGA.
#' 
#' @return Matrix containing LRR values.
#' 
#' @export
#' @import modeest minfi Biobase
#' 
#' @examples
#' 
#' \dontrun{
#' # Get LRR
#' normix <- tsh$Sample_Group == 'Normal'
#' # Median is used for quick calculation
#' test <- getLRR(tset, Normals = normix, Ref = 'median')
#' 
#' # Segment data with 4 cores
#' test.tum <- test[,!normix]
#' test.seg <- LRRtoCNA(LRR = test.tum, ncores = 4L)
#' head(test.seg$output)}
#'
#' @seealso \code{\link[minfi]{preprocessFunnorm}} 
#'    \code{\link[minfi]{RGChannelSet-class}}
#'

getLRR <- function(rgSet, Normals = NULL, sampNames = NULL, QN = FALSE, 
                   Ref = "mode", mode.bw = 0.1, mode.method = "naive", 
                   normal.cnv = TRUE, mean.center = TRUE, 
                   filterProbes = FALSE, retainedProbes = NULL,
                   keep_fnobj = FALSE, fn_output = NULL, ...) {
  
  # Checks
  
  if (!inherits(rgSet, "RGChannelSet")) {
    stop("rgset has to be RGChannelSet.")
  }
  if (!inherits(Normals, c("numeric", "logical", "RGChannelSet", "NULL", 
                           "character"))) {
    stop("Normals in wrong format.")
  }
  if (!Ref %in% c("mode", "median")) {
    stop("Ref has to be 'mode' or 'median'.")
  }
  
  # Get set for processing depending on Normal input
  
  pData(rgSet) <- .coerce.pData(pData(rgSet))
  
  if(!is.null(Normals) & length(Normals) == 1) if(is.na(Normals)){
    cat('Normals specified as NA. Reference will calculated using all user provided samples.\n')
    Normals <- rep(TRUE, ncol(rgSet))
    normal.cnv <- TRUE
  }
  
  if (inherits(Normals, c("numeric", "logical"))) {
    rgset <- rgSet
    Normals <- Normals
  }
  
  if (inherits(Normals, "RGChannelSet")) {
    if (nrow(rgSet) != nrow(Normals)) 
      stop("Features in normal and tumor set do not match.")
    rgset <- BiocGenerics::combine(rgSet, Normals)
    Normals <- colnames(rgset) %in% colnames(Normals)
  }
  
  if (is.null(Normals)) {
    library(EpicopyData)
    cat("Using all epicopy normals as reference set.\n")
    if (!"all.normals" %in% ls(globalenv())) 
      data("allNormals")
    rgset <- BiocGenerics::combine(rgSet, all.normals)
    Normals <- colnames(rgset) %in% colnames(all.normals)
  }
  
  if (inherits(Normals, "character")) {
    if (!Normals %in% c("all", "thyroid", "breast", "lung")) {
      stop("Character normals have to be one of 'all', 'thyroid', 'breast', or 'lung'.")
    }
    .Normals <- Normals
    if (Normals == "all"){
      library(EpicopyData)
      if(!"all.normals" %in% ls(globalenv())) data('allNormals')
      Normals <- all.normals
    } else
    if (Normals == "thyroid"){
      library(EpicopyData)
      if(!"all.normals" %in% ls(globalenv())) data('allNormals')
      Normals <- all.normals[,pData(all.normals)$Tissue_Type %in% 'thyroid']
    } else
    if (Normals == "breast") {
      library(EpicopyData)
      if(!"all.normals" %in% ls(globalenv())) data('allNormals')
      Normals <- all.normals[,pData(all.normals)$Tissue_Type %in% 'breast']
    } else
    if (Normals == "lung") {
      library(EpicopyData)
      if(!"all.normals" %in% ls(globalenv())) data('allNormals')
      Normals <- all.normals[,pData(all.normals)$Tissue_Type %in% 'lung']
    }
    cat("Using", .Normals, "normals as reference set.\n")
    pData(Normals) <- .coerce.pData(pData(Normals))
    rgset <- BiocGenerics::combine(rgSet, Normals)
    Normals <- colnames(rgset) %in% colnames(Normals)
  }
  
  Normals <- as.logical(Normals)
  if (any(is.na(Normals))) {
    stop("Error in classifying normals. Please review input.")
  }
  
  # Set sample names
  
  if (is.null(sampNames)) {
    colnames(rgset) <- pData(rgset)$Sample_Name
  } else {
    if (ncol(rgset) == length(sampNames)) 
      stop("sampNames is not equal to number of samples in rgset.\n")
    colnames(rgset) <- sampNames
  }
  
  # Funnorm
  
  cat("Performing functional normalization.\n")
  fn.set <- .funnorm(rgset, ...)
  fn.intensity <- getUnmeth(fn.set) + getMeth(fn.set)
  fn.intensity[fn.intensity <= 0] <- 0.01
  fn.intensity <- log2(fn.intensity)
  
  normal.fn <- fn.intensity[, Normals]
  
  if(keep_fnobj) {
    cat('Saving funnorm object with normals to specified directory.')
    if(is.null(fn_output)) fn_output <- '.'
    fn_name <- paste0(fn_output, '/epicopy_fn.rda')
    save(fn.set, file = fn_name)
  }
  
  # Quantile normalization
  
  if (QN) {
    cat("Quantile normalization of normals.\n")
    
    # All probes have been filtered to be autosomal
    normal.t1g <- normal.fn[t1g.probes, ]
    normal.t1r <- normal.fn[t1r.probes, ]
    normal.t2 <- normal.fn[t2.probes, ]
    
    stopifnot(complete.cases(normal.t1g), complete.cases(normal.t1r), 
              complete.cases(normal.t2))
    
    normal.t1g.qn <- normalize.quantiles(normal.t1g)
    dimnames(normal.t1g.qn) <- dimnames(normal.t1g)
    normal.t1r.qn <- normalize.quantiles(normal.t1r)
    dimnames(normal.t1r.qn) <- dimnames(normal.t1r)
    normal.t2.qn <- normalize.quantiles(normal.t2)
    dimnames(normal.t2.qn) <- dimnames(normal.t2)
    normal.fn <- rbind(normal.t1g.qn, normal.t1r.qn, normal.t2.qn)
    
    if (!setequal(autosomal, rownames(normal.fn))) 
      stop("Normalization returns discordant number of probes.")
  }
  
  # Subset into autosomal probe sets
  
  cat("Filtering for autosomal probes.\n")
  normal.fn <- normal.fn[autosomal, ]
  tumor.fn <- fn.intensity[autosomal, !Normals]
  
  # Filter away SNP probes
  
  if (filterProbes) {
    if(is.null(retainedProbes)){
      cat("Filtering for TCGA probeset.")
      normal.fn <- normal.fn[tcga.probeset, ]
      tumor.fn <- tumor.fn[tcga.probeset, ]
    } else {
      if(!inherits(retainedProbes, 'character')) stop('RetainedProbes has to be of
                                                      class character.\n')
      if(!all(retainedProbes %in% rownames(tumor.fn))) stop('Some retainedProbes are
                                                            not found in dataset.\n')
      cat("Filtering for user selected probeset.")
      normal.fn <- normal.fn[retainedProbes,]
      tumor.fn <- tumor.fn[retainedProbes,]
    }
  }
  
  # Calculate reference signal
  
  cat("Calculating reference intensity using", Ref, ".\n")
  if (Ref == "mode") {
    normal.ref <- apply(normal.fn, 1, function(x) {
      mlv(x, bw = mode.bw, method = mode.method, na.rm = TRUE)$M
    })
  }
  if (Ref == "median") {
    normal.ref <- apply(normal.fn, 1, median, na.rm = TRUE)
  }
  
  # Subtract reference
  
  cat("Obtaining LRR.\n")
  tumor.lrr <- sweep(tumor.fn, 1, normal.ref, "-")
  
  if (!normal.cnv) {
    final.lrr <- tumor.lrr
  } else {
    normal.lrr <- sweep(normal.fn, 1, normal.ref, "-")
    final.lrr <- cbind(tumor.lrr, normal.lrr)
  }
  
  # Mean center data
  
  if (mean.center) {
    cat("Mean centering data.\n")
    lrr.mean <- apply(final.lrr, 2, mean)
    final.lrr <- sweep(final.lrr, 2, lrr.mean, "-")
  }
  return(final.lrr)
} 

R/funnorm.R
#' @import Biobase minfi

.funnorm = function(rgSet, nPCs = 2, sex = NULL, verbose = TRUE) {
  minfi:::.isRG(rgSet)
  rgSet <- updateObject(rgSet)
  
  if (verbose) 
    cat("[preprocessFunnorm] Mapping to genome\n")
  
  gmSet <- mapToGenome(rgSet)
  subverbose <- max(as.integer(verbose) - 1L, 0)
  
  if (verbose) 
    cat("[preprocessFunnorm] Quantile extraction\n")
  
  extractedData <- minfi:::.extractFromRGSet450k(rgSet)
  
  if (is.null(sex)) {
    gmSet <- addSex(gmSet, getSex(gmSet, cutoff = -3))
    sex <- rep(1L, length(gmSet$predictedSex))
    sex[gmSet$predictedSex == "F"] <- 2L 
  }
  rm(rgSet)
  if (verbose) 
    cat("[preprocessFunnorm] Normalization\n")
  CN <- getCN(gmSet)
  gmSet <- minfi:::.normalizeFunnorm450k(object = gmSet, extractedData = extractedData, 
                                         sex = sex, nPCs = nPCs, verbose = subverbose)
  return(gmSet)
  
}
environment(.funnorm) = asNamespace('minfi')


.coerce.pData <- function(pdat) {
  pDat <- apply(pdat, 2, as.character)
  dimnames(pDat) <- dimnames(pdat)
  pDat <- as.data.frame(pDat, stringsAsFactors = FALSE)
  return(pDat)
}

R/LRRtoCNA.R
#' Convert LRR to segmented copy number
#'
#' \code{LRRtoCNA} Converts LRR to segmented copy number using the circular
#'    binart segmentation (CBS) algorithm.
#'
#' @param LRR Matrix of LRR from \code{getLRR}
#' @param sampID Sample IDs. Default: NULL, uses \code{colnames(LRR)}
#' @param ncores Integer. Default: 1L. Define number of cores for parallel
#'    segmentation. See details.
#' @param seg.undo.splits Undo.splits argument for \code{segment} from
#'    \code{DNAcopy}. Default: \code{'sdundo'}.
#' @param seg.undo.SD undo.SD argument for \code{segment}. Default: 2.
#' @param ... Passes arguments to \code{segment}.
#' 
#' @details This function coerces the LRR file into a suitable input for
#'    the \code{CNA}, \code{smooth.CNA}, and \code{segment} algorithm
#'    provided by the \code{DNAcopy} package. Parallelization for segmentation
#'    of \code{ParDNAcopy} package is implemented when \code{ncores} >= 2L.
#'    
#' @return A segmented DNAcopy object. See \code{\link[DNAcopy]{segment}}.
#' 
#' @examples
#' # Load example data
#' data(data_vignette)
#' # Run segmentation
#' epi_cna <- LRRtoCNA(epi_lrr, seg.undo.splits = 'sdundo', seg.undo.SD = 2)
#' 
#' @import minfi DNAcopy ParDNAcopy
#' @export
#' 
#' @seealso
#'    \code{\link[DNAcopy]{CNA}}
#'    \code{\link[DNAcopy]{smooth.CNA}}
#'    \code{\link[DNAcopy]{segment}}
#'    \code{\link[ParDNAcopy]{parSegment}}
#' 

LRRtoCNA = function(LRR, sampID = NULL, ncores = 1L, 
                    seg.undo.splits = 'sdundo', seg.undo.SD = 2, ...){
  
  if(!all(rownames(LRR) %in% names(hm450))) {stop('Some probe names in LRR does not exist in 450K marker file.')}
  
  # Get marker file (map)
  lrr.map = hm450[rownames(LRR)]
  lrr.chrom = as.numeric(gsub('chr','',seqnames(lrr.map)))
  lrr.loc = lrr.map$probeTarget
  
  # Define sample names
  if(is.null(sampID)){
    sampID = colnames(LRR)
  }
  
  cat('Getting CNA object.\n')
  cna = CNA(genomdat = LRR, chrom = lrr.chrom, maploc = lrr.loc, data.type = 'logratio', sampleid = sampID)
  
  cat('Smoothing CNA object.\n')
  scna = smooth.CNA(cna)
  if(!inherits(ncores,c('numeric','integer'))) {stop('ncores must be numeric/integer')}
  
  ncores = as.integer(ncores)
  if(ncores == 1){
    cat('Running segmentation with 1 core.\n')
    seg = segment(scna, undo.splits = seg.undo.splits, undo.SD = seg.undo.SD, ...)
  } else if(ncores > 1){
    cat('Running segmentation with',ncores,'cores.\n')
    seg = parSegment(CNAobj = scna, undo.splits = seg.undo.splits, undo.SD = seg.undo.SD, njobs = ncores, ...)
  } else {
    stop('ncores is not specified correctly.')
  }
  return(seg)
}

R/export_gistic.R
#' Plot coverage of 450K across hg19
#' 
#' \code{export_gistic} Exports gistic inputs.
#' 
#' @description Parses the CNA result from \code{LRRtoCNA} and writes a
#'    segment file and marker file for input into GISTIC 2.0.
#' 
#' @param input Result from \code{LRRtoCNA}.
#' @param output Output directory.
#' @param filterbycount Logical. Recommended. Should the output segment file be 
#'    filtered for having at least \code{min_probes} number of probes in the 
#'    segment.
#' @param min_probes Number of probes to filter against.
#' @param segfile_name Name of the output segment file. NULL prints segmentation_output.
#' @param marker_file Name of the output marker file. NULL prints marker_file.
#' 
#' @details Argument \code{min_probes} restricts segments to a certain number of
#'    probes or more if \code{filterbycount} is \code{TRUE}. Writes two output files,
#'    segmentation_output and marker_file which are the .seg and marker file
#'    inputs required for GISTIC 2.0.
#'    
#' @examples
#' 
#' # Load included sample segmentation data
#' data(data_vignette)
#' # Export gistic files (defaults to local directory)
#' export_gistic(epi_seg, filterbycount = TRUE, min_probes = 50)
#'
#' @import BiocGenerics Hmisc GenomeInfoDb GenomicRanges IRanges
#' @export export_gistic
#' 
#' @seealso \code{\link[minfi]{preprocessFunnorm}} 
#'    \code{\link[minfi]{RGChannelSet-class}}

export_gistic <- function(input, output_dir, filterbycount = TRUE,
                          min_probes = 50, segfile_name = NULL,
                          markerfile_name = NULL){
  probeset <- Epicopy:::hm450[rownames(input$data),]
  marker_file <- data.frame(Probe_ID = names(probeset),
                            Chromosome = as.character(seqnames(probeset)),
                            Position = probeset$probeTarget)
  
  segfile <- input$output
  colnames(segfile) <- c('ID', 'Chromosome', 'Start', 'End', 'Probe_count', 'Segment_Mean')
  segfile$Chromosome <- paste0('chr', segfile$Chromosome)
  
  if(filterbycount) segfile <- subset(segfile, Probe_count >= min_probes)
  
  # Write .seg file
  if(is.null(segfile_name)) segfile_name <- 'segmentation_output.txt'
  segfile_name <- paste0(output_dir, '/', segfile_name)
  write.table(segfile, file = segfile_name,
              quote = FALSE, sep = '\t', row.names = FALSE)
  cat('Segmentation file written.\n')
  
  # Write marker file
  if(is.null(markerfile_name)) markerfile_name <- 'epicopy_markers.txt'
  markerfile_name <- paste0(output_dir, '/', markerfile_name)
  if(markerfile_name == segfile_name) {
    cat('Marker file and seg file names are equal. 
        Changing marker file to default name.\n')
    markerfile_name <- 'epicopy_markers.txt'
    markerfile_name <- paste0(output_dir, markerfile_name)
  }
  write.table(marker_file, file = markerfile_name,
              quote = FALSE, sep = '\t', row.names = FALSE)
  cat('Marker file written.\n')
}



R/plot_segments.R
#' Plot segmented files.
#' 
#' \code{epicopy} Plot resulting segments from \code{LRRtoCNA}.
#' 
#' @description Plots segmentation output for individual samples.
#' 
#' @param CNA CNA output from LRRtoCNA
#' @param which_sample Either a numeric/integer to indicate which sample to plot or
#'    a character vector with the sample name.
#' @param threshold Numeric vector indicating where to plot dotted lines to indicate
#'    threshold. Defaults at 0.15.
#' @param ylab Y-axis labels
#' @param xlab X-axis labels
#' @param ylim Y-axis limits
#' @param title Title. Defaults to sample name with Epicopy segmentation string.
#' @param seg_lwd Line width of the segments
#' @param ... Passes arguments to plot to control everything other than segment.
#' 
#' @details Plots the segmentation results from \code{LRRtoCNA}.
#' 
#' @examples
#' 
#' # Load included example data.
#' data(data_vignette)
#' # Plot
#' plot_segments(epi_seg, which_sample = 1)
#' 
#' @import BiocGenerics Hmisc GenomeInfoDb GenomicRanges IRanges org.Hs.eg.db TxDb.Hsapiens.UCSC.hg19.knownGene
#' @export plot_segments
#' 
#' @seealso \code{\link[Epicopy]{getLRR}} \code{\link[Epicopy]{LRRtoCNA}} 
#'    \code{\link[Epicopy]{export_gistic}} \code{\link[minfi]{RGChannelSet-class}}
#'    \code{\link[minfi]{read.450k.sheet}} \code{\link[minfi]{read.450k.exp}}

plot_segments <- function(CNA, which_sample, threshold = 0.15,
                          ylab = "Copy Number", xlab = 'Genomic Location (kb)',
                          ylim = c(-2,2), title = NULL,
                          seg_lwd = 2, seg_col = c('skyblue3', 'black'),
                          ...){
  
  # Set up chromosome and plot parameters
  chrLengths <- as.numeric(as.data.frame(seqinfo(
    TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene))[1:22, 1])
  chrLengths <- chrLengths/1000
  names(chrLengths) <- rownames(as.data.frame(
    seqinfo(TxDb.Hsapiens.UCSC.hg19.knownGene)))[1:22]
  
  chrLengthsRun <- cumsum(chrLengths)
  chrLengthsRun <- append(0, chrLengthsRun)
  chromAxisLoc <- rep(0, 22)
  for(i in 1:22){chromAxisLoc[i] = mean(chrLengthsRun[c(i,i+1)])}
  chrAxes <- chrLengthsRun
  chrLengthsRun <- chrLengthsRun[-23]
  names(chrLengthsRun) <- names(chrLengths)
  totalGenome <- sum(chrLengths)
  
  # Set colors
  if(length(seg_col) != 2) stop('Seg_col does not have 2 colors. Only two colors are supported at the moment.\n')
  if(!(.isColor(seg_col))) stop('A non-valid color is found in seg_col.\n')
  colset <- rep(seg_col, 11)
  names(colset) <- names(chrLengths)
  
  # Subset data for samples
  dummy_main <- CNA$output
  colnames(dummy_main) <- c('Sample', 'Chromosome', 'Start', 'End', 
                            'N_Probes', 'Segment_Mean')
  if(!grepl('chr', dummy_main$Chromosome[1])){
    dummy_main$Chromosome <- paste0('chr', dummy_main$Chromosome)
  }
  allsamples <- unique(dummy_main$Sample)
  if(inherits(which_sample, c('numeric', 'integer'))){
    which_sample <- allsamples[which_sample]
  } else
    if(inherits(which_sample, c('character'))){
      if(!any(which_sample %in% allsamples)) 
        stop('Sample name not found in sample list.\n')
    } else {
    stop('Which_sample argument not recognized. See ?plot_segments')
  }
    
  dummy <- subset(dummy_main, Sample %in% which_sample)
  dummy$chrLengths <- chrLengthsRun[as.vector(dummy$Chromosome)]
  dummy$startAdj <- with(dummy, Start/1000 + chrLengths)
  dummy$endAdj <- with(dummy, End/1000 + chrLengths)
  
  if(is.null(title)){
    title <- paste0(unique(dummy$Sample), "\nEpicopy segmentation")
  }
  
  plot(0, xlim = c(0, totalGenome), 
       ylim = ylim,
       type = "n", 
       ylab = ylab,
       xlab = xlab)
  abline(h = 0, col = "grey35", lwd = 1.5)
  abline(v = chrAxes, lty = 2, col = "red")
  axis(side = 3, at = chromAxisLoc, tick = FALSE, labels = c(1:22), cex.axis = 0.5, 
       line= -1)
  with(dummy,
       segments(x0 = startAdj, x1 = endAdj, y0 = Segment_Mean, lwd = seg_lwd,
                col = colset[Chromosome]))
  title(main = title)
  abline(h = c(threshold, -threshold), lwd = 1.5, col = 'grey', lty = 2)
}

.isColor <- function(x)
{
  res <- try(col2rgb(x),silent=TRUE)
  return(!"try-error"%in%class(res))
}

.makecolors <- function (object, pal = cbbPalette, keep.order = TRUE) 
{
  require(RColorBrewer)
  if (is.null(dim(object))) {
    ANS = list()
    CLfactor = inherits(object, "factor")
    if (keep.order & CLfactor) {
      dum.order = levels(object)
    }
    object = as.character(object)
    tempPal = pal
    obj.unique = unique(object)
    obj.colors = tempPal[1:length(obj.unique)]
    names(obj.colors) = obj.unique
    color.set = obj.colors[object]
    names(color.set) = object
    if (keep.order & CLfactor) {
      obj.colors = obj.colors[dum.order]
    }
    ANS$Legend = obj.colors
    ANS$Colors = color.set
    return(ANS)
  }
  if (is.null(dim(object)) == FALSE) {
    color.set.mat = matrix(NA, nrow(object), ncol(object))
    legend.list = list()
    tempPal = pal
    for (i in 1:ncol(object)) {
      object.dum = apply(object, 2, as.character)
      obj.unique = unique(object.dum[, i])
      obj.colors = tempPal[1:length(obj.unique)]
      names(obj.colors) = obj.unique
      color.set.mat[, i] = obj.colors[object.dum[, i]]
      legend.list[[i]] = obj.colors
    }
    colnames(color.set.mat) = colnames(object)
    rownames(color.set.mat) = rownames(object)
    names(legend.list) = colnames(object)
    ANS = list(Legend = legend.list, Colors = color.set.mat)
    return(ANS)
  }
}


R/Epicopy_package.R
#' CNV from methylation microarray
#' 
#' @name Epicopy
#' @usage NULL
#' @aliases 'Epicopy package' Epicopy Epicopy-package
#' @docType package
#' @title Obtain segmented CNV information from Illumina Human Methylation
#'    450K Microarray
#' @description Obtain segmented CNV from Illumina 450K methylation microarrays. Currently,
#'    circular binary segmentation (CBS) is supported for segmentation. Includes
#'    methods for plotting and calculating coverage of specific genomic
#'    regions.
#' @details \tabular{ll}{
#'    Package: \tab Epicopy\cr
#'    Type: \tab Package\cr
#'    Version: \tab 0.0.0.900\cr
#'    Date: \tab 2015-02-25\cr
#'    License: \tab GPL-3\cr
#'  }
#' 
#' @author 
#' Sean Cho
#' 
#' Hyunseok Kim
#' 
#' Leslie Cope
#' 
#' Maintainer: Sean Cho; \email{sean.cho@@jhmi.edu}
#'
NULL

R/epicopy.R
#' Run Epicopy.
#' 
#' \code{epicopy} Master function to run epicopy from raw Illumina Methylation 450K data.
#' 
#' @description Parses the CNA result from \code{LRRtoCNA} and writes a
#'    segment file and marker file for input into GISTIC 2.0.
#' 
#' @param target_dir Target directory with idat files and sample sheet.
#' @param output_dir Output directory. Defaults to current directory. If FALSE,
#'    no GISTIC 2.0 files will be outputted, but can be manually obtained
#'    using the function \code{export_gistic}.
#' @param project_name Will be used as prefixes for any output.
#' @param Normals Character string identifying column in sample sheet to scan for
#'    normals. Defaults to NULL, using all included Epicopy normals. See 
#'    details for information on the normals and additional arguments.
#' @param sampNames Character string identifying column in sample sheet to use for
#'    sample names. If NULL, uses the default \code{read.450k.exp} values, which are
#'    the ChipNo_Position.
#' @param QN Logical. Quantile normalize post funnorm? Defaults to \code{FALSE}.
#' @param mode.bw If Ref = 'mode', uses this bw to estimate the mode. 
#'    Defaults to 0.1.
#' @param mode.method If Ref = 'mode', uses this method to estimate 
#'    the mode. Defaults to 'naive'.
#' @param normal.cnv Logical. Defaults to NULL which evaluates the Normal input.
#'    If Normal is user input normal tissues, then defaults to TRUE. Otherwise FALSE.
#' @param mean.center Logical. Perform mean centering of final data.
#' @param filterProbes Logical. Filter probe by TCGA defined 450K probes 
#'    adjacent to a known SNP.
#' @param retainedProbes Character. IF filterProbes is TRUE, denotes probes to keep. 
#'    If NULL, uses probes retained by TCGA.
#' @param ncores Integer. Number of cores to use for segmentation. Defaults to 1.
#' @param seg.undo.splits Undo.splits argument for \code{segment} from
#'    \code{DNAcopy}. Default: \code{'sdundo'}.
#' @param seg.undo.SD undo.SD argument for \code{segment}. Default: 2.
#'  @param filterbycount Logical. Recommended. Should the output segment file be 
#'    filtered for having at least \code{min_probes} number of probes in the 
#'    segment.
#' @param min_probes Number of probes to filter against.
#' @param verbose Logical. Verbose?
#' @param ... Passes argument to \code{getLRR}.
#' 
#' @details Epicopy reads the sample sheet provided by the user in the \code{target_dir},
#' imports the experiment, and returns a CNA file while writing gistic outputs at
#' the \code{output_dir}.
#' 
#' If the user has their own normals, specify the column name in the samplesheet
#' that contains the identifier 'normal' (not case sensitive). Else, they can
#' specify one of 'all' (default, NULL), 'breast', 'lung', or 'thyroid' to use Epicopy 
#' included normals. The last alternative is NA which uses all the samples as 
#' references, under the assumption that the \code{mode}/\code{median} of all the
#' samples are at 2n. The last is recommended either in samples with low CNV, or
#' when there are many samples. See also \code{\link[Epicopy]{getLRR}}.
#' 
#' @import BiocGenerics Hmisc GenomeInfoDb GenomicRanges IRanges modeest minfi
#' @export 
#' 
#' @examples
#' 
#' \dontrun{
#' # Run epicopy on included example TCGA data
#' input_loc <- system.file('extdata', 'raw_idat', package = 'EpicopyData')
#' # No output directory, returns only local R object
#' epi_seg <- epicopy(input_loc, output_dir = FALSE) 
#' }
#'
#' @seealso \code{\link[Epicopy]{getLRR}} \code{\link[Epicopy]{LRRtoCNA}} 
#'    \code{\link[Epicopy]{export_gistic}} \code{\link[minfi]{RGChannelSet-class}}
#'    \code{\link[minfi]{read.450k.sheet}} \code{\link[minfi]{read.450k.exp}}

epicopy <- function(target_dir, output_dir = NULL, project_name = NULL,
                    Normals = NULL, sampNames = NULL, QN = FALSE,
                    Ref = 'mode', mode.bw = 0.1, mode.method = "naive", 
                    normal.cnv = NULL, mean.center = TRUE, 
                    filterProbes = FALSE, retainedProbes = NULL, 
                    ncores = 1L, seg.undo.splits = 'sdundo', seg.undo.SD = 2,
                    filterbycount = TRUE, min_probes = 50, verbose = TRUE, ...){
  
  if(is.null(output_dir)){
    cat('Using current directory as output directory.\n')
    output_dir <- '.'
  }
  
  target_sheet <- read.450k.sheet(target_dir)

  if(!is.null(Normals)){
    if(!Normals %in% c('breast', 'lung', 'thyroid', 'all')){
      if(!inherits(Normals, 'character')) stop('Normals input not recognized. Please review and see ?epicopy.\n')
      cat('Looking for normal annotation under', Normals, 'column in samplesheet.\n')
      normal_index <- tolower(target_sheet[,Normals])
      if(!any(normal_index %in% 'normal'))
        stop('No normal sample found in annotation.\n')
      nnormals <- table(normal_index %in% 'normal')['TRUE']
      cat('Found', nnormals, 'normal samples.\n')
      Normals <- normal_index %in% 'normal'
      if(is.null(normal.cnv)) normal.cnv <- TRUE
    } else {
      if(is.null(normal.cnv)) normal.cnv <- FALSE
    }
  } else {
    if(is.null(normal.cnv)) normal.cnv <- FALSE
  }
  
  rgset <- read.450k.exp(targets = target_sheet, 
                         verbose = verbose)
  
  lrr <- getLRR(rgSet = rgset, Normals = Normals, sampNames = sampNames, QN = QN,
         Ref = Ref, mode.bw = mode.bw, mode.method = mode.method, 
         normal.cnv = normal.cnv, mean.center = mean.center, 
         filterProbes = filterProbes, ...)
  
  cna <- LRRtoCNA(lrr, ncores = ncores, seg.undo.splits = seg.undo.splits,
                  seg.undo.SD = seg.undo.SD)

  if(!is.null(output_dir)){
    if(output_dir == FALSE){
      return(cna)
    }
  }

  export_gistic(input = cna, output_dir = output_dir, filterbycount = filterbycount,
                min_probes = min_probes, segfile_name = project_name,
                markerfile_name = project_name)
  
  return(cna)
}